### PR TITLE
add resources

### PR DESCRIPTION
**Pull request by brramos**
- data center
- data centers
- incap rules
- site

### CODE REVIEW

File: incapsula/resource_data_center_servers.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+		return err
+	}
+
+	return nil
+}
+
+func resourceDataCenterServersDelete(d *schema.ResourceData, m interface{}) error {
+	client := m.(*Client)
+
+	err := client.DeleteDataCenterServers(d.Get("server_id").(int))
```
Comment (Techcadia) [inline incapsula/resource_data_center_servers.go]: ```serverID, _ := strconv.Atoi(d.Id())```

File: incapsula/resource_data_center_servers.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	// todo: review response
+
+	return nil
+}
+
+func resourceDataCenterServersUpdate(d *schema.ResourceData, m interface{}) error {
+	client := m.(*Client)
+
+	_, err := client.EditDataCenterServers(
+		d.Get("server_id").(int),
```
Comment (Techcadia) [inline incapsula/resource_data_center_servers.go]: ```d.Id()```

File: incapsula/resource_incap_rule.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+				}
+
+				siteID, err := strconv.Atoi(idSlice[0])
+				ruleID := idSlice[1]
+				if err != nil {
+					return nil, err
+				}
+
+				d.Set("site_id", siteID)
+				d.Set("rule_id", ruleID)
```
Comment (Techcadia) [inline incapsula/resource_incap_rule.go]: ```d.SetId(strconv.Itoa(ruleID))```

File: incapsula/resource_incap_rule.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+				Description: "todo",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"filter": {
+				Description: "todo",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"rule_id": {
```
Comment (Techcadia) [inline incapsula/resource_incap_rule.go]: rule_id can be removed as its the primary_key

File: incapsula/resource_incap_rule.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+	client := m.(*Client)
+
+	_, err := client.AddIncapRule(
+		d.Get("site_id").(int),
+		d.Get("enabled").(string),
+		d.Get("priority").(string),
+		d.Get("name").(string),
+		d.Get("action").(string),
+		d.Get("filter").(string),
+		d.Get("rule_id").(int),
```
Comment (Techcadia) [inline incapsula/resource_incap_rule.go]: ```d.Id()```

File: incapsula/resource_incap_rule.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+func resourceIncapRuleUpdate(d *schema.ResourceData, m interface{}) error {
+	return resourceIncapRuleCreate(d, m)
+}
+
+func resourceIncapRuleDelete(d *schema.ResourceData, m interface{}) error {
+	client := m.(*Client)
+
+	// Implement delete by clearing out the rule configuration
+	err := client.DeleteIncapRule(
+		d.Get("rule_id").(int),
```
Comment (Techcadia) [inline incapsula/resource_incap_rule.go]: ```d.Id()```

File: incapsula/resource_site.go [hunk 7]
```diff
@@ -60,6 +60,51 @@ func resourceSite() *schema.Resource {
...
+				Optional:    true,
+			},
+			"approver": &schema.Schema{
+				Description: "my.approver@email.com (some approver email address).",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"ignore_ssl": &schema.Schema{
+				Description: "true or empty string.",
+				Type:        schema.TypeString,
```
Comment (Techcadia) [inline incapsula/resource_site.go]:
```
							ValidateFunc: validation.StringInSlice([]string{
								"true",
								"",
							}, false),
```
https://github.com/hashicorp/terraform/blob/master/helper/validation/validation.go#L130

File: incapsula/resource_site.go [hunk 8]
```diff
@@ -60,6 +60,51 @@ func resourceSite() *schema.Resource {
 				Type:        schema.TypeString,
 				Optional:    true,
 			},
+			"site_id": &schema.Schema{
```
Comment (Techcadia) [inline incapsula/resource_site.go]: I am the one that removed `site_id` curious why you want to put it back when `d.Id()` is the `site_id`

File: incapsula/resource_site.go [hunk 9]
```diff
@@ -154,8 +199,112 @@ func resourceSiteRead(d *schema.ResourceData, m interface{}) error {
 }
 
 func resourceSiteUpdate(d *schema.ResourceData, m interface{}) error {
-	// Not implemented
-	return nil
+	client := m.(*Client)
+
+	siteID := d.Get("site_id").(int)
```
Comment (Techcadia) [inline incapsula/resource_site.go]: ```siteID := d.Id()```

File: incapsula/resource_site.go (hunk 8 above)
Comment (Techcadia) [reply incapsula/resource_site.go]: Ah I see it below being used home the other comment makes more sense.

File: incapsula/resource_data_center.go [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+	return &schema.Resource{
+		Create: resourceDataCenterCreate,
+		Read:   resourceDataCenterRead,
+		Update: resourceDataCenterUpdate,
+		Delete: resourceDataCenterDelete,
+		Importer: &schema.ResourceImporter{
+			State: func(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+				idSlice := strings.Split(d.Id(), "/")
+				if len(idSlice) != 2 || idSlice[0] == "" || idSlice[1] == "" {
+					return nil, fmt.Errorf("unexpected format of ID (%q), expected site_id/rule_id", d.Id())
```
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_data_center.go]: updating

File: incapsula/resource_data_center_servers.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+)
+
+func resourceDataCenterServers() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceDataCenterServersCreate,
+		Read:   resourceDataCenterServersRead,
+		Update: resourceDataCenterServersUpdate,
+		Delete: resourceDataCenterServersDelete,
+		Importer: &schema.ResourceImporter{
+			State: func(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
```
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_data_center_servers.go]: ty, updating

File: incapsula/resource_data_center.go [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+				}
+
+				siteID, err := strconv.Atoi(idSlice[0])
+				ruleID := idSlice[1]
+				if err != nil {
+					return nil, err
+				}
+
+				d.Set("site_id", siteID)
+				d.Set("rule_id", ruleID)
```
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_data_center.go]: updating

File: incapsula/resource_data_center.go [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+		d.Get("is_standby").(string),
+		d.Get("is_content").(string),
+	)
+
+	if err != nil {
+		return err
+	}
+
+	// Set the rule ID
+	d.SetId(d.Get("dc_id").(string))
```
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_data_center.go]: adding

File: incapsula/resource_data_center.go [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+		return err
+	}
+
+	return nil
+}
+
+func resourceDataCenterDelete(d *schema.ResourceData, m interface{}) error {
+	client := m.(*Client)
+
+	err := client.DeleteDataCenter(d.Get("dc_id").(int))
```
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_data_center.go]: updated

File: incapsula/resource_data_center_servers.go [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+				//}
+				//
+				//siteID, err := strconv.Atoi(idSlice[0])
+				//ruleID := idSlice[1]
+				//if err != nil {
+				//	return nil, err
+				//}
+				//
+				//d.Set("site_id", siteID)
+				//d.Set("rule_id", ruleID)
```
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_data_center_servers.go]: updated

File: incapsula/resource_data_center_servers.go [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+		d.Get("server_address").(string),
+		d.Get("is_standby").(string),
+	)
+
+	if err != nil {
+		return err
+	}
+
+	// Set the dc ID
+	d.SetId(d.Get("dc_id").(string))
```
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_data_center_servers.go]: updated

File: incapsula/resource_data_center_servers.go (hunk 2 above)
Comment (brramos) [reply incapsula/resource_data_center_servers.go]: updated

File: incapsula/resource_data_center_servers.go (hunk 1 above)
Comment (brramos) [reply incapsula/resource_data_center_servers.go]: updated

File: incapsula/resource_incap_rule.go (hunk 3 above)
Comment (brramos) [reply incapsula/resource_incap_rule.go]: updated

File: incapsula/resource_incap_rule.go (hunk 4 above)
Comment (brramos) [reply incapsula/resource_incap_rule.go]: removed

File: incapsula/resource_incap_rule.go (hunk 5 above)
Comment (brramos) [reply incapsula/resource_incap_rule.go]: updated

File: incapsula/resource_incap_rule.go (hunk 6 above)
Comment (brramos) [reply incapsula/resource_incap_rule.go]: updated

**Pull request closed without merging**